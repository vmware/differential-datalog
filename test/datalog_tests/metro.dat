start;

insert Links(4, "St.Germain", "Odeon"),
insert Links(4, "Odeon", "St.Michel"),
insert Links(4, "St.Michel", "Chatelet"),
insert Links(1, "Chatelet", "Louvres"),
insert Links(1, "Louvres", "Palais Royal"),
insert Links(1, "Palais-Royal", "Tuileries"),
insert Links(1, "Tuileries", "Concorde");

commit;

dump;

