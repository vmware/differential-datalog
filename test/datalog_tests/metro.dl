typedef line = int
typedef station = string

ground relation Links(l: line, s1: station, s2: station)



relation Reach(s1: station, s2: station)


Reach(x, y) :- Links(l, x, y).
Reach(x, y) :- Links(l, y, x).
Reach(x, y):- Links(l, x, z), Reach(z, y).

relation Answer(x: station)

Answer(x) :- Reach("Odeon", x).
