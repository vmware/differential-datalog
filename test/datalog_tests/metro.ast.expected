typedef Answer = Answer{x: station}
typedef Links = Links{l: line, s1: station, s2: station}
typedef Reach = Reach{s1: station, s2: station}
extern type Set<'A>
extern type Vec<'A>
typedef line = bigint
typedef station = string
extern function __builtin_2string (x: 'X): string
extern function hash128 (x: 'X): bit<128>
extern function hash64 (x: 'X): bit<64>
extern function hex (x: 'X): string
relation Answer [Answer]
input relation Links [Links]
relation Reach [Reach]
Reach(.s1=x, .s2=y) :- Links(.l=l, .s1=x, .s2=y).
Reach(.s1=x, .s2=y) :- Links(.l=l, .s1=y, .s2=x).
Reach(.s1=x, .s2=y) :- Links(.l=l, .s1=x, .s2=z), Reach(.s1=z, .s2=y).
Answer(.x=x) :- Reach(.s1="Odeon", .s2=x).
